graph/db/sql_store.go

```diff
@@ -521,7 +522,14 @@ func (s *SQLStore) SetSourceNode(ctx context.Context,
 	node *models.Node) error {
 
 	return s.db.ExecTx(ctx, sqldb.WriteTxOpt(), func(db SQLQueries) error {
-		id, err := upsertNode(ctx, db, node)
+		// For the source node, we use a less strict upsert that allows
+		// updates even when the timestamp hasn't changed. This handles
+		// the race condition where multiple goroutines (e.g.,
+		// setSelfNode, createNewHiddenService, RPC updates) read the
+		// same old timestamp, independently increment it, and try to
+		// write concurrently. We want all parameter changes to persist,
+		// even if timestamps collide.
```

Comment:
What if we use `upsertNode` and ignore `sql.ErrNoRows` here?

Reply:
cause `upsertNode` still uses the other query which has the strict timestamp rule. For our own node, we dont want to fail to persist our own update even if the timestamp is the same as the previous one. For our own node, it actually can be the case that we persist the same update _with different info_ which is the risk